core/shared/platform/windows/win_file.c

```diff
@@ -656,13 +863,79 @@ os_close(os_file_handle handle, bool is_stdio)
     return __WASI_ESUCCESS;
 }
 
+static __wasi_errno_t
+read_data_at_offset(HANDLE handle, const struct __wasi_iovec_t *iov, int iovcnt,
+                    __wasi_filesize_t offset, size_t *nwritten)
+{
+    OVERLAPPED *read_operations =
+        BH_MALLOC((uint32_t)(sizeof(OVERLAPPED) * (uint32_t)iovcnt));
+
+    if (read_operations == NULL)
+        return __WASI_ENOMEM;
+
+    ULARGE_INTEGER query_offset = { .QuadPart = offset };
+    __wasi_errno_t error = __WASI_ESUCCESS;
+    size_t total_bytes_read = 0;
+
+    const __wasi_iovec_t *current = iov;
+    int successful_read_count = 0;
+
+    for (int i = 0; i < iovcnt; ++i, ++current) {
+        read_operations[i].Internal = 0;
+        read_operations[i].InternalHigh = 0;
+        read_operations[i].Offset = query_offset.LowPart;
+        read_operations[i].OffsetHigh = query_offset.HighPart;
+        read_operations[i].hEvent = NULL;
+
+        if (!ReadFileEx(handle, current->buf, (DWORD)current->buf_len,
+                        &read_operations[i], NULL)) {
+            DWORD win_error = GetLastError();
+            if (win_error != ERROR_IO_PENDING) {
+                error = convert_windows_error_code(win_error);
+                break;
```

Comment:
Is this an error, shall we check whether `i == iovcnt` or `error` after the for loop to check whether there was an error?

Reply:
Not sure if I'm understanding fully but even if there is an error for one of the read operations, we should still wait on the previous successful read operations. `successful_read_count` is used to make sure we only wait on the operations which were started successfully.